Feature/task per index

ryansch opened the pull request:
Our database has grown to the point where it takes a while to index the core of some of our indexes.  The default behavior of clearing the queue and indexing everything together no longer works for us.  In response to this, I've modified your gem to create a task per index and to provide locking around the delta indexer.

Where indexing a model might create a task that indexes index1, index2, and index3 deltas, now you get three tasks that each do one of those indexes.  This makes it very easy to clear duplicates.

I've written a rake task called smart_index that parses the sphinx config to find core/delta index groups.  It then will run through these and lock the delta indexer while the core is running.  This is needed because the core index will reset the delta flag to 0 and if the delta indexer automatically runs during that time then whatever records were in the delta are now unsearchable until the core finishes.

I haven't gone so far as to toggle this behavior with a configuration value.  I wanted to get some feedback from you if you thought it would be useful to others.

ryansch committed:
Enqueuing a task per index.

ryansch committed:
Adding redis locking for indexing.

agibralter commented:
Ahh this is great timing actually. I was just thinking about the problem of the delta indexes running while the main indexer runs. But I think you're mixing up two things here...

ts-resque-delta only adds and removes from a Resque queue that processes the _delta_ indexes. And it doesn't clear the whole queue... it just clears the queue of multiple instances of the same delta queue.

For example, if you have an index on the `users` table and two new users get added to the system (let's say... quickly), two Resque jobs with the args "user_delta" are added to the queue. Now it doesn't make sense to run these one after the other because the first job will actually index all users with `delta=1` in the `users` table--when the second job runs, there won't be anything to index. This line in `delta_job.rb` removes identical jobs from the queue:

```
Resque.redis.lrem("queue:#{@queue}", 0, redis_job_value)
```

Now, I agree that the delta indexer shouldn't run while the main indexer is running. I think this would be as simple as running something like `ps aux | grep indexer`, no?

Also, are you saying that it would be good to be able to process **delta** indexes in parallel as long as they're working on different delta_cores?

ryansch commented:
I was looking at the index rake task that prepends the call to ThinkingSphinx::Deltas::ResqueDelta.cancel_thinking_sphinx_jobs when I was referring to TS clearing the queue.

As far as the duplicate removal is concerned, I think that's working perfectly!

We've got some models that update multiple indexes and those indexes can take a long time to update.  I don't want to have a blanket `ps aux | grep indexer` because there's no reason for me to disable the deltas for other indexes.  One of our core indexes takes almost an hour to update and some of the others aren't far behind.

I haven't put any thought into parallelizing any of the indexing.

Does this clear up my approach?

agibralter commented:
Ahh yeah ok. For some reason I thought you meant your delta indexing was taking a long time... hmm alright I'll take a look at the patch this evening and tomorrow and see about integration. Thank you for the pull request, btw!

agibralter commented:
(oops didn't mean to close)

ryansch commented:
You're welcome!  Let me know if you want me to refine it or add configuration/tests.
I've also published my changes as ryansch-ts-resque-delta (gem) if you'd like to play with it.

agibralter commented on lib/thinking_sphinx/deltas/resque_delta/tasks.rb:
```diff
@@ -14,9 +14,46 @@ namespace :thinking_sphinx do
+			Resque.redis.set("ts-delta:index:#{name}_delta:locked", 'true')
+			system "indexer --config #{CONFIG_FILE} --rotate #{name}_core"
+			ret = $?
+				Resque.redis.del("ts-delta:index:#{name}_delta:locked")
```
Why do you unlock the index before running the indexer again on line 47?

ryansch replied:
I'll respond to all of your patch comments first thing tomorrow morning.

On May 28, 2011, at 15:28, agibralter
reply@reply.github.com
wrote:

> > -        CONFIG_FILE ||= File.join(ENV['RAILS_ROOT'], 'config', "#{ENV['RAILS_ENV']}.sphinx.conf")
> >   +
> > -        indexes_ary = `egrep "^index\ [a-zA-Z_]+_(core|delta)" #{CONFIG_FILE} | cut -c 7- | sed -e "s/ :.*//"`
> >   +
> > -        indexes = SortedSet.new
> > -        indexes_ary.each do |index|
> > -            name, sep, type = index.rpartition('_')
> >   +
> > -            indexes << name
> > -        end
> >   +
> > -        indexes.each do |name|
> > -            Resque.redis.set("ts-delta:index:#{name}_delta:locked", 'true')
> > -            system "indexer --config #{CONFIG_FILE} --rotate #{name}_core"
> > -            ret = $?
> > -                Resque.redis.del("ts-delta:index:#{name}_delta:locked")
> 
> Why do you unlock the index before running the indexer again on line 47?
> 
> ## 
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/agibralter/ts-resque-delta/pull/5/files#r36940

agibralter replied:
Thanks!

ryansch replied:
This is the manual run of the delta indexer to make sure that it has been truncated to reflect the new state of the core index after line 43.

agibralter replied:
That makes sense, but I still don't get why you have `Resque.redis.del("ts-delta:index:#{name}_delta:locked")` before line 47...

ryansch replied:
If I were to hold the lock during the first invocation of the delta indexer, users could be adding records to the site but resque jobs wouldn't be created for those records.  Since these are being added after the delta asks the db for records to index they wouldn't be searchable until the next time the delta indexer runs.

Perhaps the cleaner solution here is to add a delta index job to the queue immediately after releasing the lock instead of running the delta manually.  This would prevent 2 delta indexers from running at the same time.

agibralter replied:
Yes, I agree. I think enqueueing the delta again makes sense. What would be the best way to do this though... what about deleted records?

``` ruby
  def index(model, instance = nil)
    return true if skip?(instance)
    Resque.enqueue(
      ThinkingSphinx::Deltas::ResqueDelta::DeltaJob,
      model.delta_index_names
    )
    if instance
      Resque.enqueue(
        ThinkingSphinx::Deltas::ResqueDelta::FlagAsDeletedJob,
        model.core_index_names,
        instance.sphinx_document_id
      )
    end
    true
  end
```

I'm about to comment on this method above... (sorry for digging into this so much -- just want to make sure all bases are covered)

agibralter replied:
Actually, I'm not -- for some reason I thought you were skipping `ThinkingSphinx::Deltas::ResqueDelta::FlagAsDeletedJob` if locked...

I think instead of line 47, it should just be:

``` ruby
Resque.enqueue(
  ThinkingSphinx::Deltas::ResqueDelta::DeltaJob,
  ["#{name}_delta"]
)
```

ryansch replied:
Done!

agibralter commented on lib/thinking_sphinx/deltas/resque_delta/tasks.rb:
```diff
@@ -14,9 +14,46 @@ namespace :thinking_sphinx do
   end
+
+	desc "Deals with large indexes one at a time with delta locking"
+	task :smart_index do
```
This could also be `task :smart_index => :environment` and we can avoid the stuff with `ENV`, right? Do you prefer being able to run this task without loading your app?

ryansch replied:
We don't need the rails environment loaded to run this task.  The ENV stuff references environment variables (not rails environment) to support the following usage:

```
rake thinking_sphinx:smart_index CONFIG_FILE="/path/to/config/file" 
```

or

```
CONFIG_FILE="/path/to/config/file" rake thinking_sphinx:smart_index
```

agibralter replied:
True, I understand `ENV["CONFIG_FILE"]` -- but you don't see `RAILS_ROOT` and `RAILS_ENV` much anymore... it's usually `Rails.root` and `Rails.env`.

ryansch replied:
That's in ruby code.  In this code here, RAILS_ROOT and RAILS_ENV are shell environment variables.  I'd much sooner remove that method of finding the config file than load the entire rails environment.

My use case is that on my servers I can autoset these environment variables with an .rvmrc and everything just works.

agibralter commented on lib/thinking_sphinx/deltas/resque_delta/tasks.rb:
```diff
@@ -14,9 +14,46 @@ namespace :thinking_sphinx do
+		end
+
+		indexes.each do |name|
+			Resque.redis.set("ts-delta:index:#{name}_delta:locked", 'true')
```
What if the delta indexer is already running here? Seems like there might be some race conditions... we may want to use a proper lock acquisition algorithm like the one used in resque-lock-timeout: https://github.com/lantins/resque-lock-timeout/blob/master/lib/resque/plugins/lock_timeout.rb

agibralter replied:
Alternatively, after the lock is acquired here, we can just check to see if `"#{name}_delta" is being indexed, and if it is we can wait until it is finished.

ryansch replied:
I don't know if we need locking here.  The condition I want to prevent is the delta indexer starting anytime after the core index starts and updates the delta flag on the records in the DB.  If the delta is already running then presumably it won't see the smaller set of delta flagged records after core updates.  I'd be ok with sleeping here after taking the lock and waiting for ps | grep to tell us that the delta indexer isn't running just to be sure.

Thoughts?

agibralter commented on lib/thinking_sphinx/deltas/resque_delta.rb:
```diff
@@ -41,10 +41,13 @@ class ThinkingSphinx::Deltas::ResqueDelta < ThinkingSphinx::Deltas::DefaultDelta
-      ThinkingSphinx::Deltas::ResqueDelta::DeltaJob,
-      model.delta_index_names
-    )
+    model.delta_index_names.each do |delta|
```
Why check for locking before enqueuing and before performing?

ryansch replied:
I know that the smart_index task is going to run the delta indexer after the core index finishes.  The lock is only held while the core is indexed so it doesn't make sense to add delta index jobs to the queue when I'm going to manually invoke the delta indexer anyway.

ryansch committed:
Enqueueing delta job in smart_index.

This is instead of manually running the delta indexer and running the risk of having 2 delta indexers running simultaneously for the same index.

ryansch commented:
As per your comment on digging in: I think it's fantastic!  Concurrency and locking are not obvious and the discussion has been fun.

agibralter commented on lib/thinking_sphinx/deltas/resque_delta/tasks.rb:
```diff
@@ -14,9 +14,48 @@ namespace :thinking_sphinx do
+	task :smart_index do
+		require 'set'
+		require 'rubygems'
+		require 'resque'
```
Ok I think this is the last thing. I don't know if we should explicitly require rubygems. It's not necessary with bundler. And for people not using bundler, well can't they just require rubygems in their Rakefile?

Oh, and `SortedSet` is Ruby 1.9 only... can we use an Arrray and use `indexes.uniq!; indexes.sort!`?

agibralter replied:
And it should probably be `require 'thinking_sphinx/deltas/resque_delta'` instead of `require 'resque'`...

ryansch replied:
Originally I wrote this as a standalone ruby script before making it a rake task.  This is just a copy/paste error.  Fixed in the latest commit.

SortedSet isn't 1.9 only.  I can require and use it with 1.8.7.

agibralter replied:
Oops. True. :)

ryansch committed:
Fixing requires for smart_index task.

agibralter commented:
As an asid , I'm also going to update the README after this merge... I was thinking about putting in a recommendation that people consider a mysql index on all `detla` columns. Any thoughts?

ryansch commented:
Absolutely.  That WHERE `table`.`delta` = 0 (or 1) is going to be expensive otherwise.  I think it's done often enough to warrant the index update on write.

agibralter commented:
Arggg tabs! :) Going to fix the whitespace now...

Anyway, I think that I should change:

``` ruby
namespace :thinking_sphinx do
  task :index do
    ThinkingSphinx::Deltas::ResqueDelta.cancel_thinking_sphinx_jobs
  end
end
```

to lock all the deltas... that way even if people don't use smart_index, they won't get conflicts and strange behavior.

agibralter commented:
Err... I would have to figure out how to unlock them after `thinking_sphinx:index` is done... is there a way to wrap rake tasks? Alternatively, I can just have `ts:in` raise an exception, saying "please use `ts:si`...

ryansch commented:
Ironically one of my coworkers just convinced me to switch to spaces after 13 years of using tabs...

I like the locking idea quite a lot.  Redis makes this stuff so much easier with its atomic operations.

ryansch commented:
Yeah actually there is.  Check out the end of this blog post: http://www.dcmanges.com/blog/modifying-rake-tasks

He talks about how to use the enhance method to make things run both before and after a rake task.

agibralter commented:
Ok... so I'm just trying to follow thinking-sphinx convention as much as possible here... All the thinking-sphinx rake tasks load the app_env. And they use ThinkingSphinx's internal knowledge of the configuration file:

``` ruby
  desc "Index data for Sphinx using Thinking Sphinx's settings"
  task :index => :app_env do
    config = ThinkingSphinx::Configuration.instance
    unless ENV["INDEX_ONLY"] == "true"
      puts "Generating Configuration to #{config.config_file}"
      config.build
    end

    FileUtils.mkdir_p config.searchd_file_path
    config.controller.index :verbose => true
  end
```

Why not leverage `config.controller.index index_name, :verbose => true`?

``` ruby
config.configuration.indexes.collect { |i| i.name }
#=> ["foo_core", "foo_delta", "foo", "bar_core", "bar_delta", "bar"]
# We can just parse out the _cores and _deltas here...
```

Since it relies on Thinking-Sphinx's internal API, this seems a little more robust than separately parsing the configuration file with unix commands.

As far as loading the whole app environment... well every thinking-sphinx command already does this. Seeing as a full index is already a slow, expensive task... I don't see why loading the app_env would be that much worse...

ryansch commented:
I'm sold.  I hadn't looked at how TS was doing the existing indexing.  This evolved from a shell script I threw together to automate our indexing after I first outgrew the rake tasks for indexing/reindexing.

I'm about to step out for a bit (about an hour and a half), I'll code this up when I get back if you haven't gotten to it.

agibralter commented:
I'll give it a shot on a branch, and if you like it I'll merge it into master.

ryansch commented:
Awesome!  Just @mention me in a comment on the commit when you've pushed your branch.
